drafts/2019-01-21-java-11.adoc

```diff
@@ -0,0 +1,293 @@
+
+Here is an example of what a `module-info.java` might look like:
+[source,java]
+----
+module com.foo {
+  // Declare what packages are exported from this module
+  exports com.foo.api;
+  exports com.foo.spi to com.foo.othermodule;
+
+  // Declare a package as "open" so that other modules can perform deep reflection into it
+  // This is useful for CDI injection or JPA entity weaving
+  opens com.foo.internal.jpa.entities;
+
+  // Declare what modules this module requires
+  // By default, every module automatically requires the 'java.base' module
+  requires java.sql;
+  requires java.xml;
+
+  // Declare used and provided services
+  uses java.sql.Driver;
+  provides com.foo.api.FooService
+      with com.foo.internal.FooServiceImpl;
+}
+----
+
+The JDK itself is now organized into Java modules, and it looks something like this:
+
+image::/img/blog/blog_java11_jpms.png[]
+
+The intent is for the entire Java ecosystem to eventually organize itself into Java modules, but for now Java modules are not defined in the JavaEE/JakartaEE/MicroProfile space. *While you can technically include a Java module in your Open Liberty application, the module-info descriptor will have no effect at the moment.*
+
+==== JPMS Access Restrictions
+
+While JPMS does introduce some interesting architectural capabilities and brings a long-needed modularity system into the JDK, it does so at the expense of several popular Java technologies, namely CDI and JPA. The reason for this is that Java modules cannot access members of non-exported packages from other Java modules, and Java code may no longer perform deep reflection (i.e. things that require `setAccessible(true)`) on members of Java modules. If either of these rules are violated, an `IllegalAccessException` is thrown.
+
+Since this is such a major compatibility issue for most of the Java ecosystem, a "kill switch" (formally known as the `--illegal-access=permit` JVM option) has been introduced. It is enabled by default and allows code in unnamed modules to access and perform deep reflection on all named modules. It is important to note that eventually the kill switch will not be enabled by default, and later it will be removed entirely, so it is best to stop writing code that uses deep reflection and start migrating to alternative approaches.
+
+== Useful references for Java 11 migration
+
+While this blog post focused mainly on Java 11 migration within the context of Liberty, there are many great resources out there for generic Java 11 migration, as well as Java 11 feature highlights. To name a few:
```

Comment:
How about referencing our Migration Toolkit?  Dalia's team has many useful rules that can help detect issues with migrating to Java 11.

Reply:
The migration toolkit is for WAS Liberty, and since this blog post is on openliberty.io, Alasdair and I were initially opposed to mentioning it. As I think about this again, I suppose its OK as long as its mixed in with other java 11 migration resources.